net/graphsync_fetcher.go

```diff
@@ -229,46 +244,65 @@ func (gsf *GraphSyncFetcher) fetchBlocksRecursively(ctx context.Context, baseCid
 	return nil
 }
 
-func (gsf *GraphSyncFetcher) loadTipsetFromCids(ctx context.Context, cids []cid.Cid) (types.TipSet, error) {
-	blks := make([]blocks.Block, 0, len(cids))
-	for _, cid := range cids {
-		block, err := gsf.store.Get(cid)
-		if err != nil {
-			return types.UndefTipSet, err
-		}
-		blks = append(blks, block)
-	}
-	validatedBlocks, err := sanitizeBlocks(ctx, blks, gsf.validator)
+// Loads the IPLD blocks for all blocks in a tipset, and checks for the presence of the
+// message and receipt list structures in the store.
+// Returns the tipset with an empty `incomplete`, or Undef and the CIDs of all blockchain-blocks
+// for which the expected IPLD blocks are not all present.
+func (gsf *GraphSyncFetcher) loadAndVerify(ctx context.Context, key types.TipSetKey) (types.TipSet, []cid.Cid, error) {
+	// Load the block headers that exist.
+	tip, incomplete, err := gsf.loadTipHeaders(ctx, key)
 	if err != nil {
-		return types.UndefTipSet, err
+		return types.UndefTipSet, nil, err
 	}
 
-	return types.NewTipSet(validatedBlocks...)
+	// Check that nested structures are also stored, recording any that are missing as incomplete.
+	for i := 0; i < tip.Len(); i++ {
+		blk := tip.At(i)
+		for _, link := range []cid.Cid{blk.Messages, blk.MessageReceipts} {
+			// TODO: validate the structures, don't just check for their presence #3232
+			ok, err := gsf.store.Has(link)
+			if err != nil {
+				return types.UndefTipSet, nil, err
+			}
+			if !ok {
+				incomplete = append(incomplete, blk.Cid())
+			}
+		}
+	}
```

Comment:
That verification could be considered redundant if we trusted graphsync to be correct and that we were using it properly - but in this case having it in the first place would have immediately exposed that we *weren't* using it properly. Since we should also validate the syntax of those other structures before accepting them, I don't think checking for them is wasteful work in the long run.